packages/ifpack2/src/Ifpack2_PowerMethod.hpp

```diff
@@ -152,70 +156,142 @@ powerMethodWithInitGuess (const OperatorType& A,
+    for(int iter = 0; iter < numIters-1; ++iter) {
+      if(verbose) {
+        *out << "  Iteration " << iter << endl;
+      }
+      A.apply(*x, *y);
+      x->elementWiseMultiply(STS::one(), D_inv, *y, STS::zero());
+
+      if(((iter+1) % eigNormalizationFreq == 0) && (iter < numIters-2)) {
+        norm = x->norm2();
+        if(norm == zero) { // Return something reasonable.
+          if(verbose) {
+            *out << "   norm is zero; returning zero" << endl;
+            *out << "   Power method terminated after "<< iter << " iterations." << endl;
+          }
+          return zero;
+        } else {
+          lambdaMaxOld = lambdaMax;
+          lambdaMax = pow(norm, Teuchos::ScalarTraits<MT>::one() / eigNormalizationFreq);
+          if(Teuchos::ScalarTraits<ST>::magnitude(lambdaMax-lambdaMaxOld) < tolerance * Teuchos::ScalarTraits<ST>::magnitude(lambdaMax)) {
+            if(verbose) {
+              *out << "  lambdaMax: " << lambdaMax << endl;
+              *out << "  Power method terminated after "<< iter << " iterations." << endl;
+            }
+            return lambdaMax;
+          } else if(verbose) {
+            *out << "  lambdaMaxOld: " << lambdaMaxOld << endl;
+            *out << "  lambdaMax: " << lambdaMax << endl;
+            *out << "  |lambdaMax-lambdaMaxOld|/|lambdaMax|: " << Teuchos::ScalarTraits<ST>::magnitude(lambdaMax-lambdaMaxOld)/Teuchos::ScalarTraits<ST>::magnitude(lambdaMax) << endl;
+          }
         }
-        return zero;
-      } else {
-        lambdaMaxOld = lambdaMax;
-        lambdaMax = pow(norm, Teuchos::ScalarTraits<MT>::one() / eigNormalizationFreq);
-        if (Teuchos::ScalarTraits<ST>::magnitude(lambdaMax-lambdaMaxOld) < tolerance * Teuchos::ScalarTraits<ST>::magnitude(lambdaMax)) {
-          if (verbose) {
+        x->scale(one / norm);
+      }
+    }
+    if(verbose) {
+      *out << "  lambdaMax: " << lambdaMax << endl;
+    }
+
+    norm = x->norm2();
+    if(norm == zero) { // Return something reasonable.
+      if(verbose) {
+        *out << "   norm is zero; returning zero" << endl;
+        *out << "   Power method terminated after "<< numIters << " iterations." << endl;
+      }
+      return zero;
+    }
+    x->scale(one / norm);
+    A.apply(*x, *y);
+    y->elementWiseMultiply(STS::one(), D_inv, *y, STS::zero());
+    lambdaMax = y->norm2();
+  } else {
+    // In this route, the update is as follows:
+    // y=A*x, y = Dinv*y, lambda = x->dot(y), x = y/lambda
+    if(verbose) {
+      *out << "  PowerMethod using largest eigenvalue route" << endl;
+    }
+    for (int iter = 0; iter < numIters-1; ++iter) {
+      if(verbose) {
+        *out << "  Iteration " << iter << endl;
+      }
+      A.apply(*x, *y);
+      y->elementWiseMultiply(STS::one(), D_inv, *y, STS::zero());
+
+      if(((iter+1) % eigNormalizationFreq == 0) && (iter < numIters-2)) {
+        norm = x->dot(*y);
```

Comment:
@GrahamBenHarper While you're debugging, might as well change that variable name.

Reply:
updated to `xDinvAx` :)